samwise.js

```diff
@@ -3,52 +3,53 @@
  *
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?